### PR TITLE
test(integration): add SSH key E2E validation

### PR DESCRIPTION
**Pull request by duplico**
## Summary

- Adds integration test fixture with RSA + ED25519 SSH keys
- Validates keys are actually installed in `/home/vyos/.ssh/authorized_keys`
- Tests multiple key handling and format preservation

## Test plan

- [x] `just check` passes (475 tests)
- [x] New fixture added to test-mapping.yml
- [ ] Integration CI validates SSH key injection works E2E

Generated with Claude Code (Opus 4.5)

### CODE REVIEW

File: tests/integration/contexts/ssh-keys.env [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# Primary IP for management/SSH
+ETH0_IP="192.168.122.50"
+ETH0_MASK="255.255.255.0"
+ETH0_GATEWAY="192.168.122.1"
+
+# SSH public keys for the vyos user
+# Testing with multiple keys (newline-separated)
+# Using real SSH key formats (RSA, ED25519) to ensure proper handling
+SSH_PUBLIC_KEY="ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDKj8HhT6Z8vN4Jx7rWqZ0V9jH1mP9qL2xF3wK5nB7sT8cY4uV6dE2fG3hI9jK0lM1nO2pQ3rS4tU5vW6xY7zA8bC9dE0fG1hI2jK3lM4nO5pQ6rS7tU8vW9xY0zA1bC2 test-key-1
+ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOMqqnkVzrm0SdG6UOoqKLsabgH5C9okWi0dh2l9GKJl test-key-2"
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/integration/contexts/ssh-keys.env]: Fixed in 0cca103bf703955baaf5d4c07d2658e3bf236e3f.

The SshKeyGenerator now splits on newlines first to handle multiple keys. Each key is parsed individually and all valid keys generate their corresponding VyOS configuration commands. The SSH service is only enabled if at least one valid key is processed.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: tests/test_ssh_integration.py [hunk 2]
```diff
@@ -146,3 +146,154 @@ def test_static_routes_if_configured(
...
+        key_lines = [line for line in output.split('\n') if line.strip().startswith('ssh-')]
+
+        # If we have multiple keys, verify they're all present
+        # The ssh-keys.env fixture has 2 keys (one RSA, one ED25519)
+        # Other fixtures may have 0 or 1 key
+        if len(key_lines) >= 2:
+            # Verify we have both RSA and ED25519 keys
+            key_types = [line.split()[0] for line in key_lines if line.strip()]
+            assert (
+                "ssh-rsa" in key_types or "ssh-ed25519" in key_types
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/test_ssh_integration.py]: Fixed in 0cca103bf703955baaf5d4c07d2658e3bf236e3f.

Changed the assertion from OR to AND to properly verify that both ssh-rsa and ssh-ed25519 key types are present when multiple keys are configured.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: tests/test_ssh_integration.py [hunk 3]
```diff
@@ -146,3 +146,154 @@ def test_static_routes_if_configured(
...
+
+        # Should have SSH service configured
+        assert "service ssh" in output
+
+        # Should be on port 22 (standard SSH port)
+        port_output = ssh_connection(
+            "show configuration | grep 'service ssh port' || echo 'default'"
+        )
+        # Either explicitly set to 22 or using default
+        assert "22" in port_output or "default" in port_output
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/test_ssh_integration.py]: Fixed in 0cca103bf703955baaf5d4c07d2658e3bf236e3f.

Added the same skip guard pattern used in other tests in the class. The test now checks for the authorized_keys file existence and skips with pytest.skip() if SSH keys are not configured.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/vyos_onecontext/generators/system.py [hunk 4]
```diff
@@ -41,37 +41,53 @@ def generate(self) -> list[str]:
...
 
-        # Use comment as identifier if available, otherwise use "key1"
-        key_id = parts[2] if len(parts) >= 3 else "key1"
-        # Sanitize key_id for use as VyOS identifier
-        # VyOS only accepts alphanumeric characters and underscores
-        # Strip surrounding quotes (single or double) that may be in the comment
-        key_id = key_id.strip("\"'")
-        key_id = key_id.replace("@", "_at_").replace(" ", "_").replace(".", "_")
+            # Use comment as identifier if available, otherwise use "key1"
+            key_id = parts[2] if len(parts) >= 3 else "key1"
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/vyos_onecontext/generators/system.py]: Fixed in a4955334e54f39b092e636dde9abef851bd97c7b. Now tracking processed keys with a counter (key_counter) to generate unique identifiers (key1, key2, key3...) for keys without comments. This prevents overwriting when multiple keys lack comments.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: tests/integration/run-qemu-test.sh [hunk 5]
```diff
@@ -476,6 +476,27 @@ case "$CONTEXT_NAME" in
...
+        # SSH key configuration commands
+        assert_command_generated "set service ssh port 22" "SSH service enabled on port 22"
+        assert_command_generated "set system login user vyos authentication public-keys" "SSH public key configuration"
+        # Verify key type and key data are configured
+        assert_command_generated "authentication public-keys.*type" "SSH key type configured"
+        assert_command_generated "authentication public-keys.*key" "SSH key data configured"
+
+        # Verify multiple keys if configured (ssh-keys.env has 2 keys)
+        # Count the number of public-key commands
+        KEY_COUNT=$(grep -c "VYOS_CMD:.*authentication public-keys.*key.*AAAA" "$SERIAL_LOG" || echo "0")
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/integration/run-qemu-test.sh]: Fixed in a4955334e54f39b092e636dde9abef851bd97c7b. Replaced the fragile "AAAA" pattern with robust counting of distinct key identifiers. Now extracts key IDs from authentication public-keys commands and counts unique entries.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: tests/test_ssh_integration.py [hunk 6]
```diff
@@ -146,3 +146,162 @@ def test_static_routes_if_configured(
...
+
+        # Count number of key lines (each valid key starts with ssh-)
+        key_lines = [line for line in output.split('\n') if line.strip().startswith('ssh-')]
+
+        # If we have multiple keys, verify they're all present
+        # The ssh-keys.env fixture has 2 keys (one RSA, one ED25519)
+        # Other fixtures may have 0 or 1 key
+        if len(key_lines) >= 2:
+            # Verify we have both RSA and ED25519 keys
+            key_types = [line.split()[0] for line in key_lines if line.strip()]
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/test_ssh_integration.py]: Fixed in a4955334e54f39b092e636dde9abef851bd97c7b. Removed the redundant line.strip() check. Since we already filter for lines starting with 'ssh-', they will never be empty strings after stripping.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/vyos_onecontext/generators/system.py [hunk 7]
```diff
@@ -41,37 +41,53 @@ def generate(self) -> list[str]:
...
+            # Sanitize key_id for use as VyOS identifier
+            # VyOS only accepts alphanumeric characters and underscores
+            # Strip surrounding quotes (single or double) that may be in the comment
+            key_id = key_id.strip("\"'")
+            key_id = key_id.replace("@", "_at_").replace(" ", "_").replace(".", "_")
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/vyos_onecontext/generators/system.py]: Fixed in a4955334e54f39b092e636dde9abef851bd97c7b. Implemented complete sanitization using regex to replace all non-alphanumeric characters (except underscores) with underscores. Preserved the special @ -> _at_ transformation for readability before the regex replacement.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/vyos_onecontext/generators/system.py [hunk 8]
```diff
@@ -41,37 +41,53 @@ def generate(self) -> list[str]:
...
+            key_configs.append(
+                f"set system login user vyos authentication public-keys {key_id} type {key_type}"
+            )
+
+        # Only enable SSH service if we have at least one valid key
+        if key_configs:
+            commands.append("set service ssh port 22")
+            commands.extend(key_configs)
+
+        return commands
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/vyos_onecontext/generators/system.py:113]: Fixed in a4955334e54f39b092e636dde9abef851bd97c7b. Added comprehensive unit tests to TestSshKeyGenerator class covering:
- Multiple newline-separated keys (with and without comments)
- Unique key ID generation (key1, key2, key3...)
- Single SSH service enablement for multiple keys
- All edge cases for multi-key scenarios

Tests validate both the key parsing logic and proper unique identifier generation.

(AI-generated via Claude Code w/ Sonnet 4.5)

File: tests/test_ssh_integration.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (duplico) [reply tests/test_ssh_integration.py]: Fully fixed in 235ebb8f5bfcb3d33d2f9dd7a7fbc249e0624cc9. Removed both occurrences of the redundant line.strip() check (line 252 in test_ssh_key_format_preserved and line 227 in test_multiple_ssh_keys_if_configured).

(AI-generated via Claude Code w/ Sonnet 4.5)

File: src/vyos_onecontext/generators/system.py [hunk 9]
```diff
@@ -41,37 +41,62 @@ def generate(self) -> list[str]:
...
-        key_type = parts[0]
-        key_data = parts[1]
+            # Sanitize key_id for use as VyOS identifier
+            # VyOS only accepts alphanumeric characters and underscores
+            # Strip surrounding quotes (single or double) that may be in the comment
+            key_id = key_id.strip("\"'")
+            # Replace @ with _at_ for better readability
+            key_id = key_id.replace("@", "_at_")
+            # Replace all other non-alphanumeric characters (except underscores) with underscores
+            import re
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/vyos_onecontext/generators/system.py]: Fixed in 8a24c38. Moved `import re` to the module level with other imports.

(AI-generated via Claude Code w/ Opus 4.5)

File: src/vyos_onecontext/generators/system.py [hunk 10]
```diff
@@ -41,37 +41,62 @@ def generate(self) -> list[str]:
...
 
-        # Use comment as identifier if available, otherwise use "key1"
-        key_id = parts[2] if len(parts) >= 3 else "key1"
-        # Sanitize key_id for use as VyOS identifier
-        # VyOS only accepts alphanumeric characters and underscores
-        # Strip surrounding quotes (single or double) that may be in the comment
-        key_id = key_id.strip("\"'")
-        key_id = key_id.replace("@", "_at_").replace(" ", "_").replace(".", "_")
+            # Increment counter for next key
+            key_counter += 1
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/vyos_onecontext/generators/system.py]: Fixed in 8a24c38. The counter now only increments when used for keys without comments, making the generated identifiers sequential (key1, key2, key3...).

(AI-generated via Claude Code w/ Opus 4.5)

File: src/vyos_onecontext/generators/system.py [hunk 11]
```diff
@@ -41,37 +43,62 @@ def generate(self) -> list[str]:
...
+                key_counter += 1  # Only increment when counter is used
+
+            # Sanitize key_id for use as VyOS identifier
+            # VyOS only accepts alphanumeric characters and underscores
+            # Strip surrounding quotes (single or double) that may be in the comment
+            key_id = key_id.strip("\"'")
+            # Replace @ with _at_ for better readability
+            key_id = key_id.replace("@", "_at_")
+            # Replace all other non-alphanumeric characters (except underscores) with underscores
+            key_id = re.sub(r"[^a-zA-Z0-9_]", "_", key_id)
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/vyos_onecontext/generators/system.py:90]: Fixed in 27545c6. Now tracking seen key IDs and appending `_2`, `_3`, etc. when duplicates are detected. Added unit test to verify behavior.

(AI-generated via Claude Code w/ Opus 4.5)
